linode/instancereservedipassignment/framework_resource.go

```diff
@@ -0,0 +1,294 @@
+			resp.Diagnostics.AddWarning(
+				"Reserved IP No Longer Exists",
+				fmt.Sprintf("Removing reserved IP %s from state because it no longer exists", state.ID.ValueString()),
+			)
+			resp.State.RemoveResource(ctx)
+			return
+		}
+		resp.Diagnostics.AddError(
+			"Unable to Refresh Reserved IP",
+			fmt.Sprintf("Error finding the specified Reserved IP: %s", err.Error()),
+		)
+		return
+	}
+
+	if ip == nil {
+		resp.Diagnostics.AddError("nil Pointer", "received nil pointer for reserved IP")
+		return
+	}
+
+	// Populate state with fetched data
+	resp.Diagnostics.Append(state.FlattenInstanceIP(ctx, *ip, false)...)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &state)...)
+}
+
+func (r *Resource) Update(
+	ctx context.Context,
+	req resource.UpdateRequest,
+	resp *resource.UpdateResponse,
+) {
+	var plan InstanceIPModel
+	var state InstanceIPModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &plan)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if !plan.RDNS.Equal(state.RDNS) {
+		rdns := plan.RDNS.ValueStringPointer()
+		updateOptions := linodego.IPAddressUpdateOptions{
+			RDNS: rdns,
+		}
+
+		client := r.Meta.Client
+		address := plan.Address.ValueString()
+		linodeID := plan.LinodeID.ValueInt64()
+
+		tflog.Debug(ctx, "client.UpdateIPAddress(...)", map[string]any{
+			"options": updateOptions,
+		})
+
+		ip, err := client.UpdateIPAddress(ctx, address, updateOptions)
+		if err != nil {
+			resp.Diagnostics.AddError(
+				"Failed to Update RDNS",
+				fmt.Sprintf(
+					"failed to update RDNS for instance (%d) ip (%s): %s",
+					linodeID, address, err,
+				),
+			)
+			return
+		}
+		if ip == nil {
+			resp.Diagnostics.AddError(
+				"Failed to Get Updated IP",
+				fmt.Sprintf(
+					"ip is a nil pointer after update operation for instance (%d) ip (%s): %s",
+					linodeID, address, err,
+				),
+			)
+			return
+		}
+		resp.Diagnostics.Append(plan.FlattenInstanceIP(ctx, *ip, true)...)
+		if resp.Diagnostics.HasError() {
+			return
+		}
+	}
+	plan.CopyFrom(ctx, state, true)
+
+	// Workaround for Crossplane issue where ID is not
+	// properly populated in plan
+	// See TPT-2865 for more details
+	if plan.ID.ValueString() == "" {
+		plan.ID = state.ID
+	}
+
+	resp.Diagnostics.Append(resp.State.Set(ctx, &plan)...)
+}
+
+func (r *Resource) Delete(
+	ctx context.Context,
+	req resource.DeleteRequest,
+	resp *resource.DeleteResponse,
+) {
+	var state InstanceIPModel
+
+	resp.Diagnostics.Append(req.State.Get(ctx, &state)...)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	client := r.Meta.Client
+	address := state.Address.ValueString()
+	linodeID := helper.FrameworkSafeInt64ToInt(
+		state.LinodeID.ValueInt64(),
+		&resp.Diagnostics,
+	)
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	tflog.Debug(ctx, "client.DeleteInstanceIPAddress(...)")
+	if err := client.DeleteInstanceIPAddress(ctx, linodeID, address); err != nil {
```

Comment:
I'm a bit hesitate on the design of the Delete(...) function here. It looks like we deleted this ip address instead of unassign it from a linode, right? By using this assignment resource, I'll assume user expect to unassign an IP instead of delete it. Do we have an unassign reserved ip endpoint? If not, I'd prefer to not implement it here.

Reply:
DELETE https://api.linode.com/{apiVersion}/linode/instances/{linodeId}/ips/{address} is a valid way to delete an additional address on a Linode

[API documentation](https://techdocs.akamai.com/linode-api/reference/delete-linode-ip
)
[Our documentation on it](https://bits.linode.com/ReservedIP/tests/blob/main/how-to.md#7-how-to-unassign-an-additional-reserved-address-to-a-linode)

Is there another API that you are thinking of that may also do the same operation?

Reply:
@yec-akamai In practice it looks like `client.DeleteInstanceIPAddress(...)` unassigns reserved IPs from the target instance without actually deleting the reserved IP, but I can't find any documentation to back this up.

Reply:
Yeah, sadly no documentation exists for the reserved IP bit of it because it is not a publicly released feature, and our API spec itself does not mention it because we did not need to do any code change for it. So the API works like it always used to, and does not do anything about reserved IP feature.